server_test.go

```diff
@@ -43,78 +48,107 @@ func (store *StubChannelStore) ChannelHasUser(channelName string, conn *SockChat
 
-		got := mustReadWSMessage(t, ws).Action
+		received := <-ws.MessageStash
 		want := "invalid_request_received"
-		AssertResponseAction(t, got, want)
+		AssertResponseAction(t, received.Action, want)
 	})
 
 	t.Run("can join a channel", func(t *testing.T) {
 		request := NewSocketMessage(JoinAction, ChannelRequest{Name: ChannelWithoutUser})
-		mustWriteWSMessage(t, ws, request)
+		ws.Write(t, request)
 
-		got := mustReadWSMessage(t, ws).Action
+		received := <-ws.MessageStash
 		want := "channel_joined"
-		AssertResponseAction(t, got, want)
+		AssertResponseAction(t, received.Action, want)
 	})
 
 	t.Run("can leave a channel", func(t *testing.T) {
 		request := NewSocketMessage(LeaveAction, ChannelRequest{Name: ChannelWithUser})
-		mustWriteWSMessage(t, ws, request)
+		ws.Write(t, request)
 
-		got := mustReadWSMessage(t, ws).Action
+		received := <-ws.MessageStash
 		want := "channel_left"
-		AssertResponseAction(t, got, want)
+		AssertResponseAction(t, received.Action, want)
 	})
 
 	t.Run("error if leaving a channel user are not in", func(t *testing.T) {
 		request := NewSocketMessage(LeaveAction, ChannelRequest{Name: ChannelWithoutUser})
-		mustWriteWSMessage(t, ws, request)
+		ws.Write(t, request)
 
-		got := mustReadWSMessage(t, ws).Action
+		received := <-ws.MessageStash
 		want := "invalid_request_received"
-		AssertResponseAction(t, got, want)
+		AssertResponseAction(t, received.Action, want)
 	})
 
 	t.Run("can not join a channel they are already in", func(t *testing.T) {
 		request := NewSocketMessage(JoinAction, ChannelRequest{Name: ChannelWithUser})
-		mustWriteWSMessage(t, ws, request)
+		ws.Write(t, request)
 
-		got := mustReadWSMessage(t, ws).Action
+		received := <-ws.MessageStash
 		want := "invalid_request_received"
-		AssertResponseAction(t, got, want)
+		AssertResponseAction(t, received.Action, want)
 	})
 
 	t.Run("can not send a message to a channel being outside of", func(t *testing.T) {
 		request := NewSocketMessage(SendMessageAction, MessageEvent{"foo", ChannelWithoutUser})
-		mustWriteWSMessage(t, ws, request)
+		ws.Write(t, request)
 
-		got := mustReadWSMessage(t, ws).Action
+		received := <-ws.MessageStash
 		want := "invalid_request_received"
-		AssertResponseAction(t, got, want)
+		AssertResponseAction(t, received.Action, want)
 
 	})
 
+	t.Run("unauthorized connection times out", func(t *testing.T) {
+		new_ws := NewTestWS(t, wsURL)
+		within(t, testTimeoutUnauthorized+20*time.Millisecond, func() { received := <-new_ws.MessageStash; assert.Equal(t, received.Action, "connection_timed_out") })
+	})
+
+	t.Run("connection timeout period is extended after logging in", func(t *testing.T) {
+		new_ws := NewTestWS(t, wsURL)
+		request := NewSocketMessage(LoginAction, LoginRequest{Nick: ValidUserNick, Password: ValidUserPassword})
+		new_ws.Write(t, request)
+		<-ws.MessageStash // read `login` response
+		within(t, testTimeoutUnauthorized+20*time.Millisecond, func() { assert.Empty(t, new_ws.MessageStash) })
```

Comment:
```suggestion
select {
case received := <-new_ws.MessageStash:
	if received.Action == "connection_timed_out" {
		t.Error("connection timed out", received)
	} else {
		t.Error("unexpected message received", received)
	}
case <-time.After(testTimeoutUnauthorized+20*time.Millisecond):
	return
}
```
`within` is wrongly used here because it immediately executes
```go
func() { assert.Empty(t, new_ws.MessageStash) }
```
and this test is about waiting more than `testTimeoutUnauthorized` and doing assertion only after that time passes